KeyVault.Acmebot/Providers/TransIpProvider.cs

```diff
@@ -0,0 +1,333 @@
+
+                return (Convert.ToBase64String(signature.Signature), body);
+            }
+
+            private void StoreToken(TransIPToken token)
+            {
+                var fullPath = Environment.ExpandEnvironmentVariables(@"%HOME%\.acme\transip_token.json");
+                var directoryPath = Path.GetDirectoryName(fullPath);
+
+                if (!Directory.Exists(directoryPath))
+                {
+                    Directory.CreateDirectory(directoryPath);
+                }
+
+                var json = JsonConvert.SerializeObject(token, Formatting.Indented);
+
+                File.WriteAllText(fullPath, json);
+            }
+
+            private TransIPToken LoadToken()
+            {
+                var fullPath = Environment.ExpandEnvironmentVariables(@"%HOME%\.acme\transip_token.json");
+
+                if (!File.Exists(fullPath))
+                    return null;
+
+                var json = File.ReadAllText(fullPath);
+
+                return JsonConvert.DeserializeObject<TransIPToken>(json);
+            }
+        }
+
+        private class TransIPToken
+        {
+            public string CustomerName { get; set; }
+
+            public string Token { get; set; }
+
+            public DateTimeOffset Expires { get; set; }
+
+            public bool IsValid()
+            {
+                return (!string.IsNullOrEmpty(Token)) && (Expires - DateTimeOffset.Now) > TimeSpan.FromMinutes(1);
+            }
+        }
+
+        private class TokenResponse
+        {
+            [JsonProperty("token")]
+            public string Token { get; set; }
+
+            public long GetTokenExpiration()
+            {
+                var token = Token.Split('.')[1];
+                token = token.PadRight(token.Length + (4 - token.Length % 4) % 4, '=');
+
+                var tokenBytes = Convert.FromBase64String(token);
+
+                var tokenObject = JObject.Parse(Encoding.UTF8.GetString(tokenBytes));
+
+                return tokenObject.Value<long>("exp");
+            }
+        }
+
+        private class TokenRequest
+        {
+            [JsonProperty("login")]
+            public string Login { get; set; }
+
+            [JsonProperty("nonce")]
+            public string Nonce { get; set; }
+
+            [JsonProperty("read_only")]
+            public bool ReadOnly { get; set; } = false;
+
+            [JsonProperty("expiration_time")]
+            public string ExpirationTime { get; set; } = "4 weeks";
+
+            [JsonProperty("label")]
+            public string Label { get; set; } = "KeyVault.Acmebot." + DateTime.UtcNow.ToString();
+
+            [JsonProperty("global_key")]
+            public bool GlobalKey { get; set; } = true;
+        }
+
+        private class ListDomainsResult
+        {
+            [JsonProperty("domains")]
+            public IEnumerable<Domain> Domains { get; set; }
```

Comment:
```suggestion
            public IReadOnlyList<Domain> Domains { get; set; }
```

`IReadOnlyList<T>` is preferred to avoid multiple enumerations.